src/types.py

```diff
@@ -599,164 +846,106 @@ def unify_with_args(self, concrete: "FunctionSignature") -> bool:
-        )
-        if fn.ret_type is not None:
-            fn_sig.return_type = type_from_ctype(fn.ret_type, typemap)
-        if fn.params is not None:
-            fn_sig.params = [
-                FunctionParam(
-                    name=param.name or "",
-                    type=type_from_ctype(param.type, typemap),
-                )
-                for param in fn.params
-            ]
-            fn_sig.params_known = True
-        return Type.function(fn_sig)
-    if isinstance(real_ctype, ca.TypeDecl):
-        if isinstance(real_ctype.type, (ca.Struct, ca.Union)):
-            struct = parse_struct(real_ctype.type, typemap)
-            return Type._ctype(struct.type, typemap, size=struct.size * 8)
-        names = (
-            ["int"] if isinstance(real_ctype.type, ca.Enum) else real_ctype.type.names
+@dataclass(eq=False)
+class StructDeclaration:
+    """Representation of a C struct or union"""
+
+    @dataclass(eq=False)
+    class StructField:
+        type: Type
+        offset_bits: int
+        name: str
+
+    size_bits: int
+    align_bits: int
+    tag_name: Optional[str]
+    typedef_name: Optional[str]
+    fields: List[StructField]  # sorted by `.offset_bits`
+    has_bitfields: bool
+    is_union: bool
+
+    def unify(
+        self,
+        other: "StructDeclaration",
+        *,
+        seen: Optional[Set["TypeData"]] = None,
+    ) -> bool:
+        # NB: Currently, the only structs that exist are defined from ctypes in the typemap,
+        # so for now we can use reference equality to check if two structs are compatible.
+        return self is other
+
+    def fields_at_offset(self, offset_bits: int) -> List[StructField]:
+        """Return the list of StructFields which contain the given offset (in bits)"""
+        fields = []
+        for field in self.fields:
+            # We assume fields are sorted by `offset_bits`, ascending
+            if field.offset_bits > offset_bits:
+                break
+            field_size_bits = field.type.get_size_bits()
+            assert field_size_bits is not None
+            if field.offset_bits + field_size_bits < offset_bits:
+                continue
+            fields.append(field)
+        return fields
+
+    @staticmethod
+    def from_ctype(ctype: Union[ca.Struct, ca.Union], universe: TypeUniverse) -> Type:
+        """
+        Return the Type representation of a given ctype struct or union, constructing a
+        StructDeclaration & registering it in the universe if it does not already exist.
+        """
+        struct_type = universe.get_struct_for_ctype(ctype)
+        if struct_type is not None:
+            return struct_type
+
+        struct = parse_struct(ctype, universe.typemap)
+
+        typedef_name: Optional[str] = None
+        if id(ctype) in universe.typemap.struct_typedefs:
+            typedef = universe.typemap.struct_typedefs[id(ctype)]
+            assert isinstance(typedef, ca.TypeDecl) and isinstance(
```

Comment:
we could make the type of struct_typedefs more precise to get rid of the first of these isinstance checks

Reply:
Good call.

If I make the following change, it gets rid of the first check:
```diff
-    struct_typedefs: Dict[Union[str, int], CType] = field(default_factory=dict)
+    struct_typedefs: Dict[Union[str, int], TypeDecl] = field(default_factory=dict)
```
Is there a trick I don't know to encode the `isinstance(typedef.type, ca.IdentifierType)` check?

Reply:
I don't think there is, no, other than changing the pycparser typing stubs to allow "ca.TypeDecl[ca.IdentifierType]" or something